### PR TITLE
Fixes memory leaks in axom

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- This PR is a bugfix for several memory leaks in axom as reported by `lsan`
   - I compiled the full stack w/ gcc@10.3.1 using the following flags 
   ```
   BLT_CXX_FLAGS =-g -mcmodel=medium -fsanitize=leak -fno-omit-frame-pointer -O3 -std=c++14
    ```
- Most leaks were in Axom tests and examples, but there were some in the code, which I'll point out. Notably:
   - fixes a leak in `sidre::MFEMSidreDataCollection` in cases where the data collection is supposed to deallocate the underlying mfem mesh
   - fixes a leak in axom::Array's move assignment operator. Thanks @publixsubfan for help with the solution!
   - There is still an outstanding leak in `multimat`, presumably related to `axom::Array` due to this line: https://github.com/LLNL/axom/blob/60a3add00eb5b9a261ca683c790a8513f5fe2e23/src/axom/multimat/multimat.cpp#L351
- Includes some minor refactoring

#### TODO

- [x] Update RELEASE-NOTES
- [x] Fix compilation error with gmock in Klee unit tests
- [x] Fix memory leak exposed through multimat's `SetAllocatorID()` function

### CODE REVIEW

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp
Comment (kennyweiss) [inline src/axom/sidre/core/MFEMSidreDataCollection.cpp:106]: @jamiebramwell -- this change might affect serac -- if you call this constructor.

If so, you'd need to add `dc->SetOwnData(false)` after calling the `MFEMSidreDataCollection` constructor. There are numerous examples in the sidre unit tests and examples.

File: src/axom/slic/core/Logger.cpp
Comment (kennyweiss) [inline src/axom/slic/core/Logger.cpp:205]: @bmhan12 -- this fixes a potential leak in the new `slic::addStreamToAllTags` function.

Since we cannot add tags to `slic` without attaching a `LogStream` (via the `slic::addStreamToTag()` function), it's not immediately clear to me how this function is intended to be used. Perhaps a clearer example demonstrating its intended usage would help.

**Comment by rhornung67** [comment]
Should we add the memory checks applied here to a GitLab CI pipeline?

**Comment by kennyweiss** [comment]
> Should we add the memory checks applied here to a GitLab CI pipeline?

Definitely. We already have an issue for that -- https://github.com/LLNL/axom/issues/1281 -- but we should hold off until after #1284 is merged

**Review by nselliott — APPROVED** [review]

File: src/axom/primal/operators/detail/clip_impl.hpp
Comment (agcapps) [inline src/axom/primal/operators/detail/clip_impl.hpp:331]: @kennyweiss , what does the added `{}` do?

**Review by agcapps — APPROVED** [review]
Thank you for the cleanup.

File: src/axom/primal/operators/detail/clip_impl.hpp
Comment (agcapps) [reply src/axom/primal/operators/detail/clip_impl.hpp:331]: Thanks!

File: src/axom/core/tests/core_array.hpp
Comment (kennyweiss) [inline src/axom/core/tests/core_array.hpp:2392]: Use-case modeled after a memory leak reported by LSAN on this function:
https://github.com/LLNL/Tribol/blob/5808ecc73d72a144c6e4d29d528e410fffe9ba81/src/redecomp/transfer/TransferByNodes.cpp#L227-L277

**Pull request merged**
